wheatley/row_generation/row_generator.py

```diff
@@ -62,9 +65,15 @@ def rounds(self) -> Row:
         return rounds(self.stage)
 
     @abstractmethod
-    def _gen_row(self, previous_row: Row, stroke: Stroke, index: int) -> Row:
+    def _gen_row_and_call(self, previous_row: Row, stroke: Stroke, index: int) -> RowCall:
```

Comment:
Instead of `RowCall` could this be modelled
```python
def get_call(self) -> Optional[Call]:
    return self._gen_call(self._index) # Maybe row/stroke if they would be needed?

#Overrde this in complib gen
def _gen_call(self, index: int) -> Optional[Call]:
    return None
```

Then bot's start next row would be something like
```python
self._row = self.row_generator.next_row()
if self.is_calling:
    call = self.row_generator.get_call()
    if call is not None:
        self.tower.make_call(call)
```

Reply:
Another alternative instead of get_call would be to add
```python
def next_row_with_call(self, make_call:Func[str]): => Row
    self._make_call(self._index, make_call)
    return self.next_row()
```

and then call it
```python
if self.is_calling:
    self._row = self.row_generator.next_row_with_call(self.tower.make_call)
else:
    self._row = self.row_generator.next_row()
```
Which would give the row generator freedom to make calls as needed without needing a reference to tower

Reply:
I don't think that having `RowGenerator`s be able to directly make calls is a good idea.  It just doesn't feel right that a part of the code titled 'row generation' should be responsible for deciding whether or not Wheatley is conducting or not.  Also, from an implementation point of view, it's really unclear what should be done in the case of having multiple calls on the same row - calling them both when the row starts is not an option because the calls would be called simultaneously, and it seems more reasonable for a `RowGenerator` to just pass control back to `Bot`, which then handles spreading the calls out.

However, I do agree that having `gen_row_and_call` almost always return `(<some row>, None)` is pretty ugly, but I like how `RowGenerator`s act like iterators do in Rust (i.e. you have a `next` method that can mutate the Iterator and returns the next value in the sequence).  I think this model is very intuitive and makes it easy to define how `Bot` should interact with the row gen module.  But I'm open to ideas - this PR is in early stages of development anyway.

To fix the returning `None`, I would suggest that we implement `_gen_row_and_calls` on the base `RowGenerator` to be implemented in terms of the abstract `_gen_row` method:
```python
def _gen_row_and_calls(self, ...):
    return (self._gen_row(...), None)
```

Then the `CompLibRowGenerator` overrides this with its own implementation and leaves `_gen_row` with some place-holder method that throws an exception when called, whilst the other `RowGenerators` override `_gen_row` and leave `_gen_row_and_call` untouched.  This way the `RowGenerator` classes that don't call anything have no added complexity, but we keep the clean interface.

What do you think?

Reply:
Yes, I'm not particularly keen on my second suggestion, though it would give the most flexibility.

To me conceptually a call doesn't relate to the row being returned, but is actually 2 rows later.
Hypothetically I'm imagining also a touch generator (being fed `WHW x3` and putting in the bobs) or a composition generator (which is whirling away in the background to come up with something that's true and comes around)

My preferred way at the moment still feels something like 
```python
def get_calls(self) -> List[str]:
    return self._gen_call(self._row, self._stroke, self._index)

def _gen_calls(self, current_row: Row, stroke: Stroke, index: int) -> List[str]:
    if index == 0 and stroke.is_hand():
        return [Calls.GO, "First method name"]
    return []



if self.is_calling:
    for call in self.row_generator.get_calls()
        #Sanitise and make call/chat
```

Which 
1) Enables this to be called during the last rounds for a bob at the very start / call of first method name
1) Is more flexible, e.g. add a `get_go_calls` instead
1) Bot's that aren't calling don't need calls anyway
1) Looks likely to be the least amount of code

If all implementers are deteriministic (i.e. `get_calls()` returns the same result until `next_row()` is called) this still feels like a senisble iteration.

For implementing a `That's all` do we actually need a peek type call? Or just accept it will be a bit late and come at the start of rounds?

The api could be changed to a different generator pattern of MoveNext / Current. I feel like I've previously wanted to get the current row multiple times, but can't recall why...
```python
def move_next() -> bool #True if it advanced the iteration (e.g. false at the end of Complib rows) 
    pass

def get_row():
    pass
def get_calls():
    pass
```

Reply:
I like MoveNext/Current pattern the most.

However, we still need a way to easily handle things like `Go` and `That's All`, which should be emitted by all row generators.  Also we need to be able to put calls on the rows before we go into changes (e.g. for [this comp](https://complib.org/composition/71359)), which I can't immediately think of a good way of doing.

Reply:
Perhaps a `def get_go_calls(stroke: Stroke)` that can be called in the final 2 rows of rounds to allow
1) Backstroke or handstroke starts (delay the call of Go to backstroke, and return rounds for the first `get_row()`)
2) Bobs at hand or back for the first or second change (I don't know an example of the latter but seems plausible)

`That's all` is presumably the penultimate loaded Complib row or the first instance of rounds as the next row for the others? 
Two options spring to mind
1) Some kind of `_peek_row()` it can use only when it is calling (e.g. load that to a `self._next_row` field from `get_calls` as `_gen_row` can't be called multiple times). In general I don't think `PlaceNotationGenerator` could peek without affecting how it reacts to bobs/singles when a user is calling.
2) Accept it will be a bit late and just have it at the start of rounds like many humans do.